feat: expand scenarios covered by default arguments for environments

shadowspawn opened the pull request:
## Problem

We are proposing to determine appropriate default arguments for `node --eval`, Electron, and ES6.

Using `require.main` to determine `--eval` (or REPL) gives false positive for ES6 main program.

Related: #5 #18

## Solution

Implement a placeholder for `process.mainArgs` that covers the scenarios that have been raised.

- detect `node --eval` directly by checking options passed to node (`-e`, `--eval`)
- detect `node --print` directly by checking options passed to node (`-p`, `--print`)
- no need to detect REPL because no user arguments from CLI?
- test for bundled Electron app (using same approach as Yargs and Commander)
- fallback to usual `process.argv.slice(2)`

Related PRs for Electron argument processing:
- https://github.com/tj/commander.js/pull/1172/files
- https://github.com/yargs/yargs/pull/1554/files

And for historical interest, the addition of `defaultApp` in Electron to work-around the variant behaviour:
- https://github.com/electron/electron/issues/4690#issuecomment-217435222

## Example

```sh
% node -e 'console.log(require(".").parseArgs())' 1 --foo 2 3
{
  args: { foo: true },
  values: { foo: [ undefined ] },
  positionals: [ '1', '2', '3' ]
}
```


```sh
 % node es6.mjs --foo 1 2 3
{
  args: { foo: true },
  values: { foo: [ undefined ] },
  positionals: [ '1', '2', '3' ]
}
```

shadowspawn commented:
Just spotted `node -p` in a comment. I had a feeling I was missing another special mode!

https://github.com/nodejs/node/pull/35015#discussion_r491625849

ljharb commented:
`node -pe` is critical

shadowspawn commented:
Added `-p` and `--print`. e.g.
```
% node --print 'require (".").parseArgs()' first second
{ args: {}, values: {}, positionals: [ 'first', 'second' ] }
```

bcoe commented on index.js:
```diff
@@ -1,7 +1,30 @@
+  // 2) process.defaultApp is either set by electron in an electron
+  //    unbundled app, or undefined
+  //    see https://github.com/electron/electron/blob/master/docs/api/process.md#processversionselectron-readonly
+  if (process.versions && process.versions.electron && !process.defaultApp) {
```
I could imagine getting some push back on this when we merge into Node.js, because it's so specific to one technology bundling Node.js.

electron is so popular I would argue on the behalf of this feature though.

shadowspawn replied:
Good comment about push back. Similarly, I was hesitant about adding Electron support to Commander to work around an historical Electron mistake.

I mentioned `process.mainArgs` in the PR. Actually, `process.mainArgs` could cover Electron too, since Electron can implement it to be aware of the Electron launch scenarios.

I'll refactor `defaultArgv` to be a placeholder for `process.mainArgs ` to make it clearer that the functionality is wanted but may need to be part of `parseArgs`.

shadowspawn commented:
Added tests, refactored code to be an explicit placeholder for `process.mainArgs`, describe Election as an example rather than as intended for inclusion.
